utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,155 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::CFStringGetCStringPtr,
+};
+use std::ffi::CStr;
+
+use crate::RetrievalError;
+
+pub fn get_locales() -> Result<Vec<String>, RetrievalError> {
+    let mut languages: Vec<String> = Vec::new();
+
+    // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
+    // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
+    // Immutablility ensures that nothing is overriden during it's scope
+    let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
+
+    if !locale_carr_ref.is_null() {
+        // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
+        let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
+        for i in 0..count {
+            // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
+            // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
+            let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref as _, i) };
+
+            if !lang_ptr.is_null() {
+                // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
+                // Returns pointer in O(1) without any memory allocation. This can return NULL so to handle it
+                let lang_str = unsafe { CFStringGetCStringPtr(lang_ptr as _, 0) };
+
+                if !lang_str.is_null() {
+                    // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
+                    let lang_rust_str_res: Result<&str, std::str::Utf8Error> =
+                        unsafe { CStr::from_ptr(lang_str) }.to_str();
+                    match lang_rust_str_res {
+                        Ok(lang_rust_str) => {
+                            languages.push(lang_rust_str.to_string());
+                        }
+                        Err(e) => {
+                            return Err(RetrievalError::ConversionError(e));
```

Comment:
you can do the same here - add `From<Utf8Error> for RetrievalError` and then you can use `?` to propagate errors.

Reply:
Makes sense now to ensure consistency 👍